src/htmlrenderer.cpp

```diff
@@ -482,6 +490,102 @@ void HtmlRenderer::render(std::istream& input,
 				}
 				break;
 			}
+
+			case HtmlTag::VIDEO: {
+				std::string videourl;
+				try {
+					videourl = xpp.get_attribute_value("src");
+				} catch (const std::invalid_argument&) {
+					videourl = "";
+				}
+				video_count++;
+				if (videourl.length() > 0) {
+					link_num = add_link(links,
+							utils::censor_url(
+								utils::absolute_url(
+									url,
+									videourl)),
+							LinkType::VIDEO);
+					curline.append(strprintf::fmt(
+							"[%s %u (%s #%u)]",
+							_("video"),
+							video_count,
+							_("link"),
+							link_num));
+					source_count++;
+				}
+				inside_video++;
+			}
+			break;
+
+			case HtmlTag::AUDIO: {
+				std::string audiourl;
+				try {
+					audiourl = xpp.get_attribute_value("src");
+				} catch (const std::invalid_argument&) {
+					audiourl = "";
+				}
+				audio_count++;
+				if (audiourl.length() > 0) {
+					link_num = add_link(links,
+							utils::censor_url(
+								utils::absolute_url(
+									url,
+									audiourl)),
+							LinkType::AUDIO);
+					curline.append(strprintf::fmt(
+							"[%s %u (%s #%u)]",
+							_("audio"),
+							audio_count,
+							_("link"),
+							link_num));
+					source_count++;
+				}
+				inside_audio++;
+			}
+			break;
+
+			case HtmlTag::SOURCE: {
+				std::string sourceurl;
+				try {
+					sourceurl = xpp.get_attribute_value("src");
+				} catch (const std::invalid_argument&) {
+					LOG(Level::WARN,
+						"HtmlRenderer::render: found "
+						"source "
+						"tag with no src attribute");
+					sourceurl = "";
+				}
+				if (inside_video && sourceurl.length() > 0) {
+					link_num = add_link(links,
+							utils::censor_url(
+								utils::absolute_url(
+									url,
+									sourceurl)),
+							LinkType::VIDEO);
+					curline.append(strprintf::fmt(
+							"[%s %u (%s #%u)]",
+							_("video"),
+							video_count,
+							_("link"),
+							link_num));
+				}
+				if (inside_audio && sourceurl.length() > 0) {
+					link_num = add_link(links,
+							utils::censor_url(
+								utils::absolute_url(
+									url,
+									sourceurl)),
+							LinkType::AUDIO);
+					curline.append(strprintf::fmt(
+							"[%s %u (%s #%u)]",
+							_("audio"),
+							audio_count,
+							_("link"),
+							link_num));
+				}
+				source_count++;
```

Comment:
The increment should happen only if the `<source>` tag has a valid source.

Reply:
It'd be nice to have a test for that :) Eventually we'll rewrite this in Rust (#89), so we better have a test suite that catches the likely bugs.

Reply:
This wasn't addressed, was it? We still don't have a test that shows that empty tags don't increase the count.

Reply:
I could extend the test for tags without valid sources to include tags with some valid sources. I'm sure that the code does the right thing (because `HtmlRenderer::add_media_link` does nothing if the link is an empty string).

Reply:
Such a test would be great; thanks!